File: 17.hs
import Data.Map qualified as M
import Data.Set qualified as S
import Control.Arrow ((&&&))

main :: IO ()
main = interact $ (++ "\n") . show . (p1 &&& p2) . parse

type Node = (Int, Int)
data Grid a = Grid { items :: M.Map Node a, lastNode :: Node }

parse :: String -> Grid Int
parse s = Grid { items = M.fromList xs, lastNode = fst (last xs) }
  where xs = [((x, y), read [c]) | (y, l) <- enum (lines s), (x, c) <- enum l]

enum :: [a] -> [(Int, a)]
enum = zip [0..]

data Direction = L | R | U | D deriving (Eq, Ord)
data Cell = Cell { node :: Node, direction :: Direction } deriving  (Eq, Ord)
data Neighbour = Neighbour { cell :: Cell, distance :: Int }

neighbours :: Grid Int -> [Int] -> Cell -> [Neighbour]
neighbours Grid { items } range = concat . adjacent
  where
    adjacent Cell { node = (x, y), direction = d }
      | d `elem` [L, R] = [cells (\m -> Cell (x, y - m) U),
                           cells (\m -> Cell (x, y + m) D)]
      | otherwise = [cells (\m -> Cell (x - m, y) R),
                     cells (\m -> Cell (x + m, y) L)]
    cells mkCell = snd $ foldl (mkNeighbour mkCell) (0, []) [1..maximum range]
    mkNeighbour mkCell (d, xs) m = let cell = mkCell m in case M.lookup (node cell) items of
      Just d2 -> (d + d2, if m `elem` range then Neighbour cell (d + d2) : xs else xs)
      _ -> (d, xs)

shortestPath :: [Int] -> Grid Int -> Int
shortestPath moveRange grid@Grid { lastNode } = go startDist S.empty startQ
  where
    -- Start in both directions so that we never have to go straight and can
    -- just always turn. This way, we don't even need to track moves.
    startCells = [Cell { node = (0, 0), direction = L },
                  Cell { node = (0, 0), direction = D }]
    startDist = M.fromList $ zip startCells [0, 0]
    startQ = S.fromList $ zip [0, 0] startCells
    isEnd Cell { node } = node == lastNode

    go ds seen q = case S.minView q of
      Nothing -> 0
      Just ((du, u), q')
        | isEnd u -> du
        | S.member u seen -> go ds seen q'
        | otherwise -> let adj = neighbours grid moveRange u
                           (ds', q'') = foldl (relax u du) (ds, q') adj
                       in go ds' (S.insert u seen) q''

    relax u du (ds, q) Neighbour { cell = v, distance = d } =
      let d' = du + d in case M.lookup v ds of
        Just dv | dv < d' -> (ds, q)
        _ -> (M.insert v d' ds, S.insert (d', v) q)

p1, p2 :: Grid Int -> Int
p1 = shortestPath [1..3]
p2 = shortestPath [4..10]

File: 17.swift
struct Grid {
    let items: [Node: Int]
    let lastNode: Node
}

struct Node: Hashable {
    let x, y : Int
}

func readInput() -> Grid {
    var items = [Node: Int]()
    var y = 0, x = 0
    while let line = readLine() {
        x = 0
        for c in line {
            items[Node(x: x, y: y)] = c.wholeNumberValue!
            x += 1
        }
        y += 1
    }
    return Grid(items: items, lastNode: .init(x: x - 1, y: y - 1))
}

enum Direction {
    case l, r, u, d
}

struct Cell: Hashable {
    let node: Node
    let direction: Direction
    let moves: Int
}

func shortestPath(grid: Grid, moveRange: ClosedRange<Int>) -> Int {
    // Starting with 0 moves allows us to consider both left and down neighbours
    // equally.
    let startCell = Cell(node: Node(x: 0, y: 0), direction: .l, moves: 0)
    func isEnd(_ cell: Cell) -> Bool { cell.node == grid.lastNode }

    func adj(_ u: Cell) -> any Sequence<Neighbour> {
        neighbours(grid: grid, moveRange: moveRange, cell: u)
    }

    var dist: [Cell: Int] = [startCell: 0]
    var seen: Set<Cell> = Set()
    var invDist: [Int: [Cell]] = [0: [startCell]]
    var minDist = 0

    func popNearest() -> (Cell, Int)? {
        // Use an inverse distance map to simulate a priority queue.
        guard minDist < invDist.count else { return nil }
        if var unseen = invDist[minDist]?.filter({ !seen.contains($0) }),
            let v = unseen.popLast() {
            let d = minDist
            if unseen.count == 0 { minDist += 1 }
            return (v, d)
        } else {
            minDist += 1
            return popNearest()
        }
    }

    while let (u, du) = popNearest() {
        if !seen.insert(u).inserted { continue }
        if isEnd(u) { return du }
        for (v, d) in adj(u) {
            let d2 = du + d
            let dv = dist[v] ?? .max
            if d2 < dv {
                dist[v] = d2
                invDist[d2, default: []].append(v)
            }
        }
    }

    return 0
}

typealias Neighbour = (cell: Cell, d: Int)

func neighbours(
    grid: Grid, moveRange: ClosedRange<Int>, cell: Cell
) -> any Sequence<Neighbour> {
    let node = cell.node
    let moves = cell.moves
    let (x, y) = (node.x, node.y)

    func make(_ xy: (Int, Int), _ direction: Direction, _ moves: Int) -> Cell {
        Cell(node: Node(x: xy.0, y: xy.1), direction: direction, moves: moves)
    }

    let cellRange = 1...moveRange.upperBound

    func seq(_ makeCell: (Int) -> Cell) -> [Neighbour] {
        var s = 0
        return Array(cellRange.compactMap { m in
            let cell = makeCell(m)
            guard let d = grid.items[cell.node] else { return nil }
            s += d
            return moveRange.contains(cell.moves) ? (cell: cell, d: s) : nil
        })
    }

    switch cell.direction {
    case .l:
        return [seq({ m in make((x + m, y), .l, moves + m) }),
                seq({ m in make((x, y - m), .u, m) }),
                seq({ m in make((x, y + m), .d, m) })].joined()
    case .r:
        return [seq({ m in make((x - m, y), .r, moves + m) }),
                seq({ m in make((x, y - m), .u, m) }),
                seq({ m in make((x, y + m), .d, m) })].joined()
    case .u:
        return [seq({ m in make((x, y - m), .u, moves + m) }),
                seq({ m in make((x - m, y), .r, m) }),
                seq({ m in make((x + m, y), .l, m) })].joined()
    case .d:
        return [seq({ m in make((x, y + m), .d, moves + m) }),
                seq({ m in make((x - m, y), .r, m) }),
                seq({ m in make((x + m, y), .l, m) })].joined()
    }
}

let grid = readInput()
let p1 = shortestPath(grid: grid, moveRange: 1...3)
let p2 = shortestPath(grid: grid, moveRange: 4...10)
print(p1, p2)
